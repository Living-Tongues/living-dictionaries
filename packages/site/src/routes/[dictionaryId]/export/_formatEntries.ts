import type { IDictionary, IEntry, ISpeaker } from '@living-dictionaries/types';
import { glossingLanguages } from './_glossing-languages-temp';
import { semanticDomains } from '$lib/mappings/semantic-domains';
import { partsOfSpeech } from '$lib/mappings/parts-of-speech';
import { friendlyName } from '$lib/helpers/friendlyName';
import { replaceHTMLTags } from '$lib/export/replaceHTMLTags';

function turnArrayIntoPipedString(itemsFormatted, i, values, columnName, fn) {
  if (values) {
    let stringValue = '';
    // There are some dictionaries (e.g. Kalanga) that have strings as sources instead of arrays
    values = typeof values === 'string' ? [values] : values;
    const list = values.map(fn);
    //In case some strings contain commas
    stringValue += list.map((el) => el.replace(/,/g, ' -'));
    stringValue = stringValue.replace(/,/g, ' | ');
    itemsFormatted[i][columnName] = stringValue;
  } else {
    itemsFormatted[i][columnName] = '';
  }
}

export function formatEntriesForCSV(
  data: IEntry[],
  { name: dictionaryName, glossLanguages }: IDictionary,
  speakers: ISpeaker[]
) {
  //Getting the total number of semantic domains by entry if they have at least one
  let totalSDN = 0;
  const filterSDN = data.filter((entry) => (entry.sdn ? entry.sdn.length : ''));
  if (filterSDN.length > 0) {
    totalSDN = filterSDN
      .map((entry) => entry.sdn.length)
      .reduce((maxLength, sdnLength) => Math.max(maxLength, sdnLength));
  }

  const replacementChars = {
    ',': ' -',
    '"': "'",
  };

  const headers = {
    id: 'Entry id',
    lx: 'Lexeme/Word/Phrase',
    ph: 'Phonetic (IPA)',
    in: 'Interlinearization',
    mr: 'Morphology',
    plf: 'Plural form',
    di: 'Dialect for this entry',
    nt: 'Notes',
    psab: 'Parts of speech abbreviation',
    ps: 'Parts of speech',
    sr: 'Source(s)',
  };

  //Assigning semantic domains as headers
  if (totalSDN > 0) {
    for (let index = 0; index < totalSDN; index++) {
      headers[`sd${index + 1}`] = `Semantic domain ${index + 1}`;
    }
  }

  //Assigning gloss languages as gloss headers
  glossLanguages.forEach((bcp) => {
    headers[`gl${bcp}`] = `${glossingLanguages[bcp]} Gloss`;
  });

  //Assigning vernacular and gloss languages as example sentence headers
  headers['xsvn'] = `Example sentence in ${dictionaryName}`;
  glossLanguages.forEach((bcp) => {
    headers[`xs${bcp}`] = `Example sentence in ${glossingLanguages[bcp]}`;
  });

  //Assigning audio metadata as headers
  Object.assign(headers, {
    sfFriendlyName: 'Audio filename',
    sfsn: 'Speaker name',
    sfbp: 'Speaker birthplace',
    sfde: 'Speaker decade',
    sfge: 'Speaker gender',
  });

  //Assigning images metadata as headers
  headers['pfFriendlyName'] = 'Image filename';

  const itemsFormatted = [];
  data.forEach((entry, i) => {
    // Replace null values with empty string
    const entryKeys = Object.keys(entry);
    entryKeys.forEach((key) => (!entry[key] ? (entry[key] = '') : entry[key]));

    itemsFormatted.push({
      id: entry.id,
      lx: entry.lx.replace(/[,"\r\n]/g, (m) => replacementChars[m]),
      ph: entry.ph ? entry.ph.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      in: entry.in ? entry.in.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      mr: entry.mr ? entry.mr.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      plf: entry.plf ? entry.plf.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      di: entry.di ? entry.di.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      nt: entry.nt
        ? replaceHTMLTags(entry.nt.replace(/[,"\r\n]/g, (m) => replacementChars[m]))
        : '',
      //xv: entry.xv,
    });

    //Assigning parts of speech (abbreviation & name)
    if (entry.ps) {
      const pos = partsOfSpeech.find((ps) => ps.enAbbrev === entry.ps)?.enName;
      if (!pos) {
        Object.assign(
          itemsFormatted[i],
          JSON.parse(`{
          "psab": "",
          "ps": "${entry.ps}"
        }`)
        );
      } else {
        Object.assign(
          itemsFormatted[i],
          JSON.parse(`{
          "psab": "${entry.ps}",
          "ps": "${pos}"
        }`)
        );
      }
    } else {
      Object.assign(itemsFormatted[i], {
        psab: '',
        ps: '',
      });
    }

    //Assigning sources
    turnArrayIntoPipedString(itemsFormatted, i, entry.sr, 'sr', (el) => el);

    //Assigning semantic domains
    for (let index = 0; index < totalSDN; index++) {
      itemsFormatted[i][`sd${index + 1}`] = '';
      if (entry.sdn && entry.sdn[index]) {
        const matchingDomain = semanticDomains.find((sd) => sd.key === entry.sdn[index]);
        if (matchingDomain) {
          itemsFormatted[i][`sd${index + 1}`] = matchingDomain.name.replace(
            /[,"\r\n]/g,
            (m) => replacementChars[m]
          );
        }
      }
    }

    //Assigning glosses
    glossLanguages.forEach((bcp) => {
      const cleanEntry = entry.gl[bcp]
        ? replaceHTMLTags(entry.gl[bcp].replace(/[,"\r\n]/g, (m) => replacementChars[m]))
        : '';
      itemsFormatted[i][`gl${bcp}`] = cleanEntry;
    });
    //Assigning example sentences
    for (let j = 0; j <= glossLanguages.length; j++) {
      if (entry.xs) {
        if (j === glossLanguages.length) {
          itemsFormatted[i][`xs${glossLanguages[j] ? glossLanguages[j] : 'vn'}`] = `${
            entry.xs['vn'] ? entry.xs['vn'].replace(/[,"\r\n]/g, (m) => replacementChars[m]) : ''
          }`;
        } else {
          itemsFormatted[i][`xs${glossLanguages[j] ? glossLanguages[j] : 'vn'}`] = `${
            entry.xs[glossLanguages[j]] ? entry.xs[glossLanguages[j]] : ''
          }`;
        }
      } else {
        itemsFormatted[i][`xs${glossLanguages[j] ? glossLanguages[j] : 'vn'}`] = '';
      }
    }
    //Audio metadata
    if (entry.sf && entry.sf.path) {
      const speaker = speakers.find((speaker) => speaker?.id === entry.sf.sp);
      let speakerName = speaker?.displayName || entry.sf.speakerName || '';
      speakerName = speakerName.replace(/[,"\r\n]/g, (m) => replacementChars[m]);
      let speakerBP = speaker?.birthplace || '';
      speakerBP = speakerBP.replace(/[,"\r\n]/g, (m) => replacementChars[m]);
      const speakerDecade = speaker?.decade || '';
      const speakerGender = speaker?.gender || '';
      Object.assign(itemsFormatted[i], {
        sfpa: entry.sf.path,
        sfFriendlyName: friendlyName(entry, entry.sf.path),
        sfsn: speakerName,
        sfbp: speakerBP,
        sfde: speakerDecade,
        sfge: speakerGender,
      });
    } else {
      Object.assign(itemsFormatted[i], {
        sfFriendlyName: '',
        sfsn: '',
        sfbp: '',
        sfde: '',
        sfge: '',
      });
    }
    if (entry.pf && entry.pf.path) {
      itemsFormatted[i]['pfpa'] = entry.pf.path;
      itemsFormatted[i]['pfFriendlyName'] = friendlyName(entry, entry.pf.path);
    } else {
      itemsFormatted[i]['pfFriendlyName'] = '';
    }
    i++;
  });
  itemsFormatted.unshift(headers);
  return itemsFormatted;
}
