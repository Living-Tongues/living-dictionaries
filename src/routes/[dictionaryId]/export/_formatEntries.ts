import type { IDictionary, IEntry } from '$lib/interfaces';
import { glossingLanguages } from './_glossing-languages-temp';
import { semanticDomains } from '$lib/mappings/semantic-domains';
import { partsOfSpeech } from '$lib/mappings/parts-of-speech';
import { fetchSpeakers } from '$lib/helpers/fetchSpeakers';
import { friendlyName } from '$lib/helpers/friendlyName';

function turnArrayIntoPipedString(itemsFormatted, i, values, columnName, fn) {
  if (values) {
    let stringValue = '';
    // There are some dictionaries (e.g. Kalanga) that have strings as sources instead of arrays
    values = typeof values === 'string' ? [values] : values;
    const list = values.map(fn);
    //In case some strings contain commas
    stringValue += list.map((el) => el.replace(/,/g, ' -'));
    stringValue = stringValue.replace(/,/g, ' | ');
    Object.assign(itemsFormatted[i], JSON.parse(`{"${columnName}": "${stringValue}"}`));
  } else {
    Object.assign(itemsFormatted[i], JSON.parse(`{ "${columnName}": "" }`));
  }
}

export async function formatEntriesForCSV(
  data: IEntry[],
  { name: dictionaryName, glossLanguages }: IDictionary
) {
  //Getting the total number of semantic domains by entry if they have at least one
  let totalSDN = 0;
  const filterSDN = data.filter((entry) => (entry.sdn ? entry.sdn.length : ''));
  if (filterSDN.length > 0) {
    totalSDN = filterSDN
      .map((entry) => entry.sdn.length)
      .reduce((maxLength, sdnLength) => Math.max(maxLength, sdnLength));
  }

  const replacementChars = {
    ',': ' -',
    '"': "'",
  };

  const speakers = await fetchSpeakers(data);

  const headers = {
    id: 'Entry id',
    lx: 'Lexeme/Word/Phrase',
    ph: 'Phonetic (IPA)',
    in: 'Interlinearization',
    mr: 'Morphology',
    di: 'Dialect for this entry',
    nt: 'Notes',
    psab: 'Parts of speech abbreviation',
    ps: 'Parts of speech',
    sr: 'Source(s)',
  };

  //Assigning semantic domains as headers
  if (totalSDN > 0) {
    for (let index = 0; index < totalSDN; index++) {
      Object.assign(headers, JSON.parse(`{"sd${index + 1}": "Semantic domain ${index + 1}"}`));
    }
  }

  //Assigning gloss languages as gloss headers
  glossLanguages.forEach((bcp) => {
    Object.assign(headers, JSON.parse(`{ "gl${bcp}": "${glossingLanguages[bcp]} Gloss" }`));
  });

  //Assigning vernacular and gloss languages as example sentence headers
  Object.assign(headers, JSON.parse(`{"xsvn": "Example sentence in ${dictionaryName}"}`));
  glossLanguages.forEach((bcp) => {
    Object.assign(
      headers,
      JSON.parse(`{"xs${bcp}": "Example sentence in ${glossingLanguages[bcp]}"}`)
    );
  });

  //Assigning audio metadata as headers
  Object.assign(headers, {
    auFriendlyName: 'Audio filename',
    ausn: 'Speaker name',
    aubp: 'Speaker birthplace',
    aude: 'Speaker decade',
    auge: 'Speaker gender',
  });

  //Assigning images metadata as headers
  Object.assign(headers, {
    imFriendlyName: 'Image filename',
  });

  const itemsFormatted = [];
  data.forEach((entry, i) => {
    // Replace null values with empty string
    const entryKeys = Object.keys(entry);
    entryKeys.forEach((key) => (!entry[key] ? (entry[key] = '') : entry[key]));

    itemsFormatted.push({
      id: entry.id,
      lx: entry.lx.replace(/[,"\r\n]/g, (m) => replacementChars[m]),
      ph: entry.ph ? entry.ph.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      in: entry.in ? entry.in.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      mr: entry.mr ? entry.mr.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      di: entry.di ? entry.di.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      nt: entry.nt ? entry.nt.replace(/[,"\r\n]/g, (m) => replacementChars[m]) : '',
      //xv: entry.xv,
    });

    //Assigning parts of speech (abbreviation & name)
    if (entry.ps) {
      const pos = partsOfSpeech.find((ps) => ps.enAbbrev === entry.ps)?.enName;
      if (!pos) {
        Object.assign(
          itemsFormatted[i],
          JSON.parse(`{
          "psab": "",
          "ps": "${entry.ps}"
        }`)
        );
      } else {
        Object.assign(
          itemsFormatted[i],
          JSON.parse(`{
          "psab": "${entry.ps}",
          "ps": "${pos}"
        }`)
        );
      }
    } else {
      Object.assign(itemsFormatted[i], {
        psab: '',
        ps: '',
      });
    }

    //Assigning sources
    turnArrayIntoPipedString(itemsFormatted, i, entry.sr, 'sr', (el) => el);

    //Assigning semantic domains
    if (entry.sdn) {
      for (let index = 0; index < totalSDN; index++) {
        Object.assign(
          itemsFormatted[i],
          JSON.parse(
            `{"sd${index + 1}": "${
              entry.sdn[index]
                ? semanticDomains
                    .find((sd) => sd.key === entry.sdn[index])
                    ?.name.replace(/[,"\r\n]/g, (m) => replacementChars[m]) || ''
                : ''
            }"}`
          )
        );
      }
    } else {
      for (let index = 0; index < totalSDN; index++) {
        Object.assign(
          itemsFormatted[i],
          JSON.parse(`{
            "sd${index + 1}": ""
          }`)
        );
      }
    }
    //Assigning glosses
    glossLanguages.forEach((bcp) => {
      const cleanEntry = entry.gl[bcp]
        ? entry.gl[bcp].replace(/[,"\r\n]/g, (m) => replacementChars[m])
        : '';
      Object.assign(itemsFormatted[i], JSON.parse(`{"gl${bcp}": "${cleanEntry}"}`));
    });
    //Assigning example sentences
    for (let j = 0; j <= glossLanguages.length; j++) {
      if (entry.xs) {
        if (j === glossLanguages.length) {
          Object.assign(
            itemsFormatted[i],
            JSON.parse(`{
              "xs${glossLanguages[j] ? glossLanguages[j] : 'vn'}": "${
              entry.xs['vn'] ? entry.xs['vn'].replace(/[,"\r\n]/g, (m) => replacementChars[m]) : ''
            }"
            }`)
          );
        } else {
          Object.assign(
            itemsFormatted[i],
            JSON.parse(`{
              "xs${glossLanguages[j] ? glossLanguages[j] : 'vn'}": "${
              entry.xs[glossLanguages[j]] ? entry.xs[glossLanguages[j]] : ''
            }"
            }`)
          );
        }
      } else {
        Object.assign(
          itemsFormatted[i],
          JSON.parse(`{
            "xs${glossLanguages[j] ? glossLanguages[j] : 'vn'}": ""
          }`)
        );
      }
    }
    //Audio metadata
    if (entry.sf && entry.sf.path) {
      const speaker = speakers.find((speaker) => speaker?.id === entry.sf.sp);
      let speakerName = speaker?.displayName || entry.sf.speakerName || '';
      speakerName = speakerName.replace(/[,"\r\n]/g, (m) => replacementChars[m]);
      let speakerBP = speaker?.birthplace || '';
      speakerBP = speakerBP.replace(/[,"\r\n]/g, (m) => replacementChars[m]);
      const speakerDecade = speaker?.decade || '';
      const speakerGender = speaker?.gender || '';
      Object.assign(itemsFormatted[i], {
        aupa: entry.sf.path,
        auFriendlyName: friendlyName(entry, entry.sf.path),
        ausn: speakerName,
        aubp: speakerBP,
        aude: speakerDecade,
        auge: speakerGender,
      });
    } else {
      Object.assign(itemsFormatted[i], {
        auFriendlyName: '',
        ausn: '',
        aubp: '',
        aude: '',
        auge: '',
      });
    }

    if (entry.pf && entry.pf.path) {
      Object.assign(itemsFormatted[i], {
        impa: entry.pf.path,
        imFriendlyName: friendlyName(entry, entry.pf.path),
      });
    } else {
      Object.assign(itemsFormatted[i], { imFriendlyName: '' });
    }
    i++;
  });
  itemsFormatted.unshift(headers);
  return itemsFormatted;
}
