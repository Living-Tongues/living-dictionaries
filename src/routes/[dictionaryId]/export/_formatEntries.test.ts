import type { IDictionary, IEntry, ISpeaker } from '$lib/interfaces';
import { formatEntriesForCSV } from './_formatEntries';

test('Example to demonstrate separate file snapshots.', () => {
  const entriesArray: IEntry[] = [
    {
      id: '12ar',
      lx: 'banana',
      lo: 'کیلا',
      ph: 'banana',
      gl: { es: 'platano' },
      ps: '1',
      sdn: ['2.1.2.3'],
      di: 'dialect x',
      nt: 'This is an example of a note, here we can write whatever we want.',
      sr: ['A book', 'www.mybook.com'],
      xs: { en: 'This is a banana' },
      sf: { path: 'https://example.com', sp: 'aa34' },
      pf: { gcs: 'https://example.com', path: 'https://example.com' },
    },
    { id: '34qw', lx: 'tree', gl: { es: 'arbol' } },
  ];
  const dictionary: IDictionary = { name: 'test', glossLanguages: ['ar', 'en'], entryCount: 0 };
  const speakers: ISpeaker[] = [{ displayName: 'John Smith', id: '123' }];
  expect(formatEntriesForCSV(entriesArray, dictionary, speakers)).toMatchSnapshot();
});
