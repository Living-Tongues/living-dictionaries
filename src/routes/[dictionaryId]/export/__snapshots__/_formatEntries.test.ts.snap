// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example to demonstrate separate file snapshots. 1`] = `
Array [
  Object {
    "di": "Dialect for this entry",
    "glar": "Arabic Gloss",
    "glen": "English Gloss",
    "id": "Entry id",
    "in": "Interlinearization",
    "lx": "Lexeme/Word/Phrase",
    "mr": "Morphology",
    "nt": "Notes",
    "pfFriendlyName": "Image filename",
    "ph": "Phonetic (IPA)",
    "ps": "Parts of speech",
    "psab": "Parts of speech abbreviation",
    "sd1": "Semantic domain 1",
    "sfFriendlyName": "Audio filename",
    "sfbp": "Speaker birthplace",
    "sfde": "Speaker decade",
    "sfge": "Speaker gender",
    "sfsn": "Speaker name",
    "sr": "Source(s)",
    "xsar": "Example sentence in Arabic",
    "xsen": "Example sentence in English",
    "xsvn": "Example sentence in test",
  },
  Object {
    "di": "dialect x",
    "glar": "",
    "glen": "",
    "id": "12ar",
    "in": "",
    "lx": "banana",
    "mr": "",
    "nt": "This is an example of a note - here we can write whatever we want.",
    "pfFriendlyName": "12ar_platano.com",
    "ph": "banana",
    "ps": "1",
    "psab": "",
    "sd1": "",
    "sfFriendlyName": "12ar_platano.com",
    "sfbp": "",
    "sfde": "",
    "sfge": "",
    "sfpa": "https://example.com",
    "sfsn": "",
    "sr": "A book | www.mybook.com",
    "xsar": "",
    "xsen": "This is a banana",
    "xsvn": "",
  },
  Object {
    "di": "",
    "glar": "",
    "glen": "",
    "id": "34qw",
    "in": "",
    "lx": "tree",
    "mr": "",
    "nt": "",
    "pfFriendlyName": "",
    "ph": "",
    "ps": "",
    "psab": "",
    "sd1": "",
    "sfFriendlyName": "",
    "sfbp": "",
    "sfde": "",
    "sfge": "",
    "sfsn": "",
    "sr": "",
    "xsar": "",
    "xsen": "",
    "xsvn": "",
  },
]
`;
